gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -346,11 +350,9 @@ def self.run_sig(hook_mod, method_name, original_method, declaration_block)
   end
 
   def self.run_builder(declaration_block)
-    builder = DeclBuilder.new(declaration_block.mod, declaration_block.raw)
-    builder
-      .instance_exec(&declaration_block.blk)
-      .finalize!
-      .decl
+    builder = declaration_block.decl_builder.deep_clone
+    builder.run!(&declaration_block.blk)
+    builder.decl
```

Comment:
I agree that this will solve the problem raised by the original test case.

I am concerned that we're now making every signature building operation more expensive by allocating more.  My questions here would be:

1. Why wasn't the newly-added test case causing problems before?  Is that just because we changed the structure of the code, since what could be delayed until the execution of the `sig`'s block now has to be executed up-front?
2. Can we get back to the original behavior by restructuring the code?

Reply:
OK, it's very possible I'm misunderstanding something, but before, we were only ever allocating `DeclBuilder` here, in `run_builder`.  Now, we're allocating *two* `DeclBuilder` objects: one in `T::Private::Methods._declare_sig_internal` when starting the `sig`, and one here, when we're actually running it.

I can see the objects for parsing and storing the currently-being-parsed `sig` state have become somewhat more complicated, but it's not immediately obvious to me why we were able to get by before *without* duplicating information here.  And that's what I'm pushing on here: does having chained `sig` syntax necessarily require us to have duplication here when we didn't before?  If so, why?  And if not, can we rewrite the code so that we don't have duplication here?  (I think it would be OK if you wound up fusing together objects or similar; I don't think we are 100% tied to the current setup of how `sig` building is done.)

(I do see your excellent description of how the added testcase illustrates the need for the cloning that you do.  I'd wonder why we don't see the same issues with the current setup, and if we could rewrite things to be closer to the current setup.)

Full disclosure: I'm not 100% sure how much the efficiency of `sig` building matters.  It is solely (at least for Stripe) a load-time thing, but we also run...many parallel tests, so load time matters more than you might think.  But I don't have concrete numbers to say "an extra allocation will slow down loading by Y%", so I'm just going on the principle of "avoid making things slower unless you have to".

Reply:
Thanks for the detailed reply @froydnj. I'm sorry I didn't understand it before, I think it was due to how similarly `DeclBuilder` and `DeclarationBlock` are named.

Creating a `DeclBuilder` in `_declare_sig_internal` is part of the core logic of this feature. We need it so that `sig` method can return a `DeclBuilder` instead of `nil`. As a result we can chain methods on it. Previously since we return `nil`, only impact of `declare_sig` was creating a `DeclarationBlock` and mark it as active. Later on we created a `DeclarationBlock` during execution. Therefore, no duplication of `DeclarationBlock` necessary. We simply didn't need that information earlier.

I'll try to see if we can solve the edge case through a different method. And if not, we might have to do a larger refactor of the feature. You are right to bring this up and personally I'd hate to slow things down or increase memory usage because we are not properly handling an edge case.